content/blog/c-off-ramp.md

```diff
@@ -0,0 +1,218 @@
+
+With Vala, I recommend making the initial translation as “light” as possible: don’t use lots of Vala’s features to rewrite code. Translate `char *` pointers to string offsets (that is, an integer string index). Make sure the translated code works first, then start simplifying. Even the “light” translation will contain simplifications, anyway: no more `malloc` or `free`, for example.
+
+### You already have a test suite, of course
+
+The other way you avoid a plague of new bugs is with a good test suite. Sorry! There’s no way around that. If you want to alter, let alone completely rewrite code, you need to have tests.
+
+Fortunately, I had already added a fairly extensive test suite to Zile years ago, when I was trying to make sure that it really behaved exactly like Emacs. The situation with Enchant was even better: it has an exhaustive (and sometimes exhausting) test suite that checks every aspect of its API, added a few years after Enchant was written, by a commission from [SIL International](https://www.sil.org). (A big shout-out to those who commission work on libre software, especially this sort of important under-the-hood work. These tests have saved me from introducing bugs time and time again.)
+
+Again, you have to be disciplined: no turning off the tests, or changing them as you go. Complete the translation without changing them, as far as possible.
+
+### Your existing build tool is probably supported
+
+Many Vala projects these days are using the [Meson](https://mesonbuild.com) build system, which is very popular in the GNOME space. But Vala is also supported well by older tools, such as [CMake](https://cmake.org) and [GNU Autotools](https://en.wikipedia.org/wiki/GNU_Autotools). Both of the projects I translated were using GNU Autotools, and I stuck with it, for two main reasons:
+
+1. I use the excellent [gnulib](https://www.gnu.org/software/gnulib/) portability library to ensure that my POSIX-compliant code runs on most systems without problems. Unfortunately, gnulib is tied to GNU Autotools. But then, few build systems work on as many platforms as the Autotools in any case.
+2. [Automake](https://www.gnu.org/software/automake/) not only has Vala support, but allows you to ship the generated C sources, so that users can build your code without a Vala compiler. This means that your translated project has exactly the same build dependencies as it had before, which can help with users and packagers who haven’t heard of Vala and don’t want to learn. With Zile, previously a pure-POSIX project, I added GLib as a build-time dependency, but of course that’s so widely used that it’s unlikely to be a problem. Enchant already used GLib.
+
```

Comment:
Unfortunately, that simply won't happen. gnulib is basically implemented *in* autotools! A more realistic prospect that I may work on myself one day is to make a conventional library, "libgnu" from gnulib. gnulib is a source library that uses autotools fragments to configure and build individual source modules inside a project; an obvious "hack" is simply to build all gnulib modules (so far as possible) into a conventional dynamic library, and then allow other projects to build against that in the normal way.

To put it another way, gnulib is basically a set of polyfills just like in the JavaScript world, only it's C + autotools rather than JavaScript + Babel + Webpack or whatever.

Reply:
My idea is to just not give a "outdated" kind of impression on the project (which is the case with autotools) so just having said that you would like to move to a more modern solution would be nice, to counter that. nothing big.

Reply:
Ah, you mean, not giving the impression that Enchant is outdated? Autotools is not outdated, by the way; it's mature, but still developed, and for certain use cases (for example, writing portable software), is still an excellent solution. I'm not too worried: users of Enchant, i.e. programmers, don't have to care about Enchant's own build system, because they will use it via their distro, and for example Enchant VAPIs are supplied "in the box" with Vala, so you can use it very easily with Vala. Maybe I should mention that?

Reply:
its not really outdated, but not at all recommended for new project anymore I would say. and I would like to make clear that valas main build system is meson and people should use that, and not autotools.

Reply:
OK, well I already said at the top of this section that most projects these days use Meson.